### PR TITLE
workflows: Add checkpatch.pl

### LINKED ISSUE #68: Add `checkpatch.pl` as GitHub action
[`checkpatch.pl`](https://docs.kernel.org/dev-tools/checkpatch.html) is a Perl script used for coding style checking in the Linux kernel. We want to use it for C source code.

Add the [`checkpatch.pl` GitHub action](https://github.com/marketplace/actions/checkpatch-pl-pr-review) to this repository.

### PR DESCRIPTION
**Pull request by RaresCroicia**
Fixes #68

Might need further changes since this is the default version of `checkpatch.pl` and it doesn't verify only the source codes.
I will look more into it, but this is the starting point.

Signed-off-by: Rares Croicia <rarescroicia@gmail.com>

### CODE REVIEW

**Review by teodutu — COMMENTED** [review]
This looks good at first sight.

@RaresCroicia our existing coding style checks are handled by [`super-linter`](https://github.com/github/super-linter). Is there a way to add the `checkpatch.pl` as another `super-linter` check?

@razvand, do you think we should add some custom scripts or arguments to this PR, or leave it for the future? I'd say we can deal with this later and get this merged sooner.

**Comment by RaresCroicia** [comment]
> Is there a way to add the `checkpatch.pl` as another `super-linter` check?

There should be, by adding the `Run checkpatch review` step in the `super-linter.yml`.
I am not sure if it works, that's why I am not pushing it yet to the PR. I will wait for an approval that this might/will work.

```yml
   .
   .
   .
   steps:
        - name: Checkout Code
          uses: actions/checkout@v3
          with:
            # Full git history is needed to get a proper list of changed files within `super-linter`
            fetch-depth: 0
  
        - name: Lint Code Base
          uses: github/super-linter@v4
          env:
            # Don't check already existent files
            VALIDATE_ALL_CODEBASE: false
            VALIDATE_GITHUB_ACTIONS: false
            VALIDATE_CLANG_FORMAT: false
            LINTER_RULES_PATH: /.github/workflows/
            MARKDOWN_CONFIG_FILE: config/config.json
            MARKDOWN_CUSTOM_RULE_GLOBS: rules/rules.js
            DEFAULT_BRANCH: master
            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       - name: Run checkpatch review
          uses: webispy/checkpatch-action@v9
```

**Comment by teodutu** [comment]
> > Is there a way to add the `checkpatch.pl` as another `super-linter` check?
> 
> There should be, by adding the `Run checkpatch review` step in `super-linter.yml`. I am not sure if it works, that's why I am not pushing it yet to the PR. I will wait for an approval that this might/will work.

It makes sense. @razvand should we merge this as it is?

**Comment by razvand** [comment]
> > > Is there a way to add the `checkpatch.pl` as another `super-linter` check?
> > 
> > 
> > There should be, by adding the `Run checkpatch review` step in `super-linter.yml`. I am not sure if it works, that's why I am not pushing it yet to the PR. I will wait for an approval that this might/will work.
> 
> It makes sense. @razvand should we merge this as it is?

It doesn't make sense to add it as a rule to `super-linter.yml` as it's not connected to `super-linter`. In the ideally case, it would be `super-linter` calling `checkpatch`. Since this is not the case, a separate `checkpatch.yml` file is the way to go.

**Comment by razvand** [comment]
> @razvand, do you think we should add some custom scripts or arguments to this PR, or leave it for the future? I'd say we can deal with this later and get this merged sooner.

Yes, let's do a minimal configuration for `checkpatch.pl` and then we can merge it. Lemme check what are good options.

**Comment by razvand** [comment]
> > @razvand, do you think we should add some custom scripts or arguments to this PR, or leave it for the future? I'd say we can deal with this later and get this merged sooner.
> 
> Yes, let's do a minimal configuration for `checkpatch.pl` and then we can merge it. Lemme check what are good options.

@RaresCroicia , I recommend we used the following configuration in [`.checkpatch.conf`](https://github.com/webispy/checkpatch-action#checkpatchpl-configuration), based on [documentation](https://docs.kernel.org/dev-tools/checkpatch.html).

```text
# This isn't a Linux kernel tree.
--no-tree

# Show the message type in the output.
--show-types
```

I wouldn't add anything else. I think the default configuration options of `checkpatch.pl` are pretty sane.

**Comment by razvand** [comment]
> it doesn't verify only the source codes

What do you mean by this?

File: .github/workflows/checkpatch.yml
Comment (razvand) [inline .github/workflows/checkpatch.yml:13]: I would add the `Info` rule from [here](https://github.com/webispy/checkpatch-action/blob/master/.github/workflows/main.yml#L14).

**Review by razvand — CHANGES_REQUESTED** [review]
Please improve the commit message.

Use `workflows: Add checkpatch.pl as GitHub action for static checking`.

The add a descriptive message (after a blank line of the title message), telling about the addition of the `checkpatch.yml` and `.checkpatch.conf` files.

At the end, at the `Fixes #68` line.

**Comment by RaresCroicia** [comment]
> > it doesn't verify only the source codes
> 
> 
> 
> What do you mean by this?

As you pointed in the issue, we want to use checkpatch for C source codes. Checkpatch verifies also the form of the commits, the messages etc., so I pointed that out.

**Comment by razvand** [comment]
> > > it doesn't verify only the source codes
> > 
> > What do you mean by this?
> 
> As you pointed in the issue, we want to use checkpatch for C source codes. Checkpatch verifies also the form of the commits, the messages etc., so I pointed that out.

I find that useful. Leave it like that. More checks, healthier codebase and commit history.

**Comment by razvand** [comment]
@RaresCroicia , be sure to credit, as part of the commit message, the source of the GitHub action config files: https://github.com/webispy/checkpatch-action

**Comment by razvand** [comment]
@RaresCroicia  , once you make updates, use the `Resolve conversation` button to mark comments as solved. And then re-request reviews: top right, in the `Reviewers` area.

Also, be sure that checks are passed. And if not, that you have reasons for them not to require being passed.

**Comment by razvand** [comment]
@RaresCroicia , you did a merge commit. I wonder why, I wonder how. Please use a single commit, use `git commit --amend` to amend an existing commit.

**Comment by razvand** [comment]
@RaresCroicia , squash the two commits, we only need one commit.

**Comment by RaresCroicia** [comment]
I messed up with a merge (I clicked sync on the wrong branch on my fork by mistake)
I reset the branch to the last one i've done, and remade the commits. 
I hope I didn't mess up again with anything.

File: .checkpatch.conf
Comment (razvand) [inline .checkpatch.conf:20]: And ending newline.

File: .checkpatch.conf
Comment (razvand) [inline .checkpatch.conf:6]:
```suggestion

# Emit diffed file position, not input file position.
--showfile

```

**Review by razvand — CHANGES_REQUESTED** [review]
See my comments. Update the commit description to:

```
Set `checkpatch.pl` GitHub action (https://github.com/webispy/checkpatch-action) for pull requests. Add `checkpatch.yml` file to configure workflow and and `.checkpatch.conf` to configure `checkpatch.pl`.
```

**Review by razvand — APPROVED** [review]
@RaresCroicia , I approved this. Before we merge it, please wrap the commit description to 72 or 80 characters. You needn't have a single long line per-sentence.

**Comment by RaresCroicia** [comment]
All done!

**Pull request merged**
